# Gemini Embedding 2 Testing Notebook
This notebook explores the multimodal embedding capabilities of `gemini-embedding-2-preview` using the `google-genai` SDK. It covers Text, Images, Video, Audio, Documents (PDFs), Interleaved Modalities, and Matryoshka Representation Learning (flexible dimensions).




[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1s2KTb9Zbhp5PdNEPSPIEHs4d-c4iaU2N?usp=sharing)

## 1. Setup
First, we'll install the required libraries and set up our API key.


In [1]:
!pip install -q google-genai pillow
import os
from google import genai
from google.genai import types
import json
import numpy as np

# NOTE: Set your GOOGLE_API_KEY environment variable before running this cell.
# Alternatively, uncomment and replace 'YOUR_API_KEY_HERE':
os.environ['GOOGLE_API_KEY'] = 'AIzaSyD0tSLoO_ZBV6eOB3V8eGxgZqWrMQ_vOrc'

client = genai.Client()
MODEL_ID = 'gemini-embedding-2-preview'


## 2. Helper Functions (Download Media)
We'll need some sample multimodal files (image, video, audio, and PDF) to test the embeddings.


In [24]:
import urllib.request

# Download an image
urllib.request.urlretrieve('https://storage.googleapis.com/generativeai-downloads/images/scones.jpg', 'scones.jpg')

# Download a video
urllib.request.urlretrieve('https://storage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4', 'reef.mp4')

# Download an audio clip
urllib.request.urlretrieve('https://storage.googleapis.com/generativeai-downloads/data/State_of_the_Union_Address_30_January_1961.mp3', 'speech.mp3')

# Download a smaller PDF (1 page) to ensure we stay within the 6-page limit
urllib.request.urlretrieve('https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf', 'report.pdf')

print("Sample files downloaded successfully!")

Sample files downloaded successfully!


## 3. Text Embeddings
Embed generic text. `gemini-embedding-2-preview` supports up to 8192 tokens of context.


In [14]:
text_query = "What are some creative ways to serve scones?"

result = client.models.embed_content(
    model=MODEL_ID,
    contents=text_query
)

[embedding] = result.embeddings
print(f"Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])


Embedding dimension: 3072
First 5 values: [0.024622202, 0.016419925, -0.004277457, -0.027584486, 0.013858194]


## 4. Image Embeddings
Embed a single image. You can process up to 6 images per request.


In [15]:
with open("scones.jpg", "rb") as f:
    image_bytes = f.read()

result = client.models.embed_content(
    model=MODEL_ID,
    contents=[
        types.Part.from_bytes(
            data=image_bytes,
            mime_type='image/jpeg',
        )
    ]
)

[embedding] = result.embeddings
print(f"Image Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])


Image Embedding dimension: 3072
First 5 values: [-0.01000852, -0.020561954, 0.0044386, 0.011214323, 0.0002687471]


## 5. Video Embeddings
Embed videos. The model supports up to 120 seconds of video input in MP4 and MOV formats.


In [16]:
with open("reef.mp4", "rb") as f:
    video_bytes = f.read()

result = client.models.embed_content(
    model=MODEL_ID,
    contents=[
        types.Part.from_bytes(
            data=video_bytes,
            mime_type='video/mp4',
        )
    ]
)

[embedding] = result.embeddings
print(f"Video Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])

Video Embedding dimension: 3072
First 5 values: [0.009374075, 0.017135378, 0.0066948906, 0.021956464, 0.005023024]


## 6. Audio Embeddings
Natively embed audio data without prior transcription.


In [18]:
with open("speech.mp3", "rb") as f:
    audio_bytes = f.read()

# gemini-embedding-2-preview handles up to 300 seconds of audio.
# We will use the first few megabytes to ensure we stay within request limits.
result = client.models.embed_content(
    model=MODEL_ID,
    contents=[
        types.Part.from_bytes(
            data=audio_bytes[:1000000], # Sending first 1MB as a sample
            mime_type='audio/mpeg', # Changed from audio/mp3 for better compatibility
        )
    ]
)

[embedding] = result.embeddings
print(f"Audio Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])

Audio Embedding dimension: 3072
First 5 values: [-0.0040408126, 0.0118788425, -0.02204797, 0.001228037, -0.0021013]


## 7. Document (PDF) Embeddings
Embed a PDF document directly (up to 6 pages).


In [25]:
with open("report.pdf", "rb") as f:
    pdf_bytes = f.read()

# Using the full PDF. gemini-embedding-2-preview handles up to 6 pages.
result = client.models.embed_content(
    model=MODEL_ID,
    contents=[
        types.Part.from_bytes(
            data=pdf_bytes,
            mime_type='application/pdf',
        )
    ]
)

# Accessing the first embedding in the list
embedding = result.embeddings[0]
print(f"PDF Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])

PDF Embedding dimension: 3072
First 5 values: [-0.023199132, -0.002439268, -0.0049567795, 0.0050121, 0.0066209687]


## 8. Interleaved Modalities
Pass multiple modalities of input (e.g., text + image) in a single request.


In [23]:
result = client.models.embed_content(
    model=MODEL_ID,
    contents=[
        "This is an image of scones:",
        types.Part.from_bytes(
            data=image_bytes,
            mime_type='image/jpeg',
        )
    ]
)

# result.embeddings contains the embedding(s).
# We take the first one which represents the combined multimodal input.
embedding = result.embeddings[0]
print(f"Interleaved (Text+Image) Embedding dimension: {len(embedding.values)}")
print("First 5 values:", embedding.values[:5])

Interleaved (Text+Image) Embedding dimension: 3072
First 5 values: [-0.0065966514, 0.013450888, -0.0016460991, -0.011051835, 0.013746156]


## 9. Flexible Output Dimensions (Matryoshka Representation Learning)
Control the embedding size manually to balance performance and storage using the `output_dimensionality` parameter. The model recommends 3072, 1536, or 768 dimensions.


In [26]:

# Generate full dimension embedding
result_full = client.models.embed_content(
    model=MODEL_ID,
    contents="Using Matryoshka Representation Learning"
)

# Generate lower dimension (768) embedding
result_low_dim = client.models.embed_content(
    model=MODEL_ID,
    contents="Using Matryoshka Representation Learning",
    config=types.EmbedContentConfig(output_dimensionality=768)
)

print(f"Default embedding length: {len(result_full.embeddings[0].values)}")
print(f"Custom scaled-down embedding length: {len(result_low_dim.embeddings[0].values)}")



Default embedding length: 3072
Custom scaled-down embedding length: 768
